sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala

```diff
@@ -218,13 +218,22 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  private def canonicalizeDeterministic(u: PythonUDF) = {
```

Comment:
The fix itself is logically making sense. But my concern is that we set this deterministic flag true by default for Python UDFs (and Scala UDFs too). When users just use UDFs actually without knowing it (which I believe arguably pretty common), they will get the same answer from UDFs users intended to work as non-deterministically after this fix.

@gatorsmile and @cloud-fan, actually, shouldn't we set it as `false` by default? - I am reading https://github.com/apache/spark/commit/ebc24a9b7fde273ee4912f9bc1c5059703f7b31e. It's an arbitrary user-defined function so I think it makes sense to have the most loose condition.

Reply:
I see your point but if we can't rely on the deterministic flag then we have serious issues in other rules as well. Please consider the example in the description. I think column `d` should be equal to `c` no matter if the UDF is deterministic or not. The `CollapseProject` rule duplicates the UDF call (because it is flagged as deterministic) and that's why we have 2 instances when `ExtractPythonUDFs` starts to run.

The following happens without this PR:
```
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CollapseProject ===
!Project [a#7, b#8, c#12, c#12 AS d#16]         Project [_1#2 AS a#7, _2#3 AS b#8, dummyUDF(_1#2) AS c#12, dummyUDF(_1#2) AS d#16]
!+- Project [a#7, b#8, dummyUDF(a#7) AS c#12]   +- LocalRelation [_1#2, _2#3]
!   +- Project [_1#2 AS a#7, _2#3 AS b#8]       
!      +- LocalRelation [_1#2, _2#3] 

...

=== Applying Rule org.apache.spark.sql.execution.python.ExtractPythonUDFs ===
!Project [_1#2 AS a#7, _2#3 AS b#8, dummyUDF(_1#2) AS c#12, dummyUDF(_1#2) AS d#16]   Project [_1#2 AS a#7, _2#3 AS b#8, pythonUDF1#22 AS c#12, pythonUDF1#22 AS d#16]
!+- LocalRelation [_1#2, _2#3]                                                        +- BatchEvalPython [dummyUDF(_1#2), dummyUDF(_1#2)], [pythonUDF0#21, pythonUDF1#22]
!                                                                                        +- LocalRelation [_1#2, _2#3]
```
which can mean different values for `c` and `d` if the UDF is non-deterministic but is flagged deterministic.

Reply:
> actually, shouldn't we set it as false by default? 

The problem is the performance regression caused by changing the default value. What we can do now is educating the users harder, like mentioning this thing in the Scala/Python UDF doc.

Reply:
Indeed, the example in the PR should have c and d equal no matter what, ideally, as it shouldn't be evaluated multiple times. But then, being deterministic doesn't matter; how would changing the default be the right fix? I don't doubt it should be 'fixed' so that even non-deterministic UDFs aren't surprisingly reevaluated. Is the point that the current handling for non-deterministic UDFs slower, even though it evaluates them just once by design? that seems weird.

Reply:
I think @cloud-fan was referring to that if we change the default to non-deterministic then it means that some of the optimization rules will not handle those UDF expressions and will leave them untouched. E.g. `PushDownPredicates` will not push them down, which can cause performance regression.

IMHO, it is the user's responsibility to set the deterministic flag right regardless what is the default. And if a UDF is flagged deterministic we should do the optimizations.

Reply:
OK I see. But surely the safer assumption is that it is not deterministic, if the user doesn't say? I get that it harms perf, but could be a correctness issue? well I trust the judgment of people here that know much more about it

Reply:
Just 2 cents, educating the users harder can be done in any way (regardless of the default value), and imagine the education doesn't work, deterministic by default would be much critical on the impact. (we can't trade off correctness and performance.) I'd in favor of consider it non-deterministic by default.

Reply:
Changing the default is orthogonal to this PR IMO.

Reply:
I agree that is orthogonal and isn't a blocker for this PR. Would we like to move the discussion further on dev@ mailing list, or it sounds like we are happy with filing a new JIRA issue and raising a PR to change the default?